File: src/android/DbCopyPlugin.java
package CommunityPlugins.DbCopy.android;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import android.util.Log;
import android.content.Context;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.FileNotFoundException;
import android.Manifest;
import android.content.pm.PackageManager;
import android.util.Base64;
import java.io.FileOutputStream;

public class DbCopyPlugin extends CordovaPlugin {
private static final int REQUEST_CODE = 123; // Any unique number for your request code


    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {

        if (action.equals("copyDbFromStorage")) {
            return copyDbFromStorage(args, callbackContext);
        } else if (action.equals("copyDbToStorage")) {
            return copyDbToStorage(args, callbackContext);
        }
        return false;
    }

    private boolean copyDbFromStorage(JSONArray args, CallbackContext callbackContext) {
        try {
            JSONObject options = args.getJSONObject(0);
            String dbName = options.getString("dbName");
            String base64Source = options.getString("sourcePath"); // Now the base64-encoded string
            String location = options.optString("location", "default");
            boolean deleteOldDb = options.optBoolean("deleteOldDb", false);

            // Determine the destination path for the database
            String destPath = getDatabasePath(location, dbName);

            // Check if the old DB exists and if it should be deleted
            File destFile = new File(destPath);
            if (deleteOldDb && destFile.exists()) {
                if (!destFile.delete()) {
                    callbackContext.error("Failed to delete old database.");
                    return false;
                }
            }

            // Decode Base64 string and write it to a temporary file
            byte[] fileBytes = Base64.decode(base64Source, Base64.DEFAULT);
            File tempFile = new File(cordova.getActivity().getCacheDir(), dbName);

            try (FileOutputStream fos = new FileOutputStream(tempFile)) {
                fos.write(fileBytes);
            }

            // Copy the temporary file (decoded from Base64) to the destination
            if (!copyFile(tempFile, destFile)) {
                callbackContext.error("Failed to copy database from the Base64 source.");
                return false;
            }

            callbackContext.success("Database copied successfully.");
            return true;
        } catch (JSONException e) {
            callbackContext.error("JSON Exception: " + e.getMessage());
            return false;
        } catch (IOException e) {
            callbackContext.error("IO Exception: " + e.getMessage());
            return false;
        }
    }

    // Helper method to get the database path based on location
    private String getDatabasePath(String location, String dbName) {
        // Assuming location can be 'default', 'documents', or 'external'
        Context context = this.cordova.getActivity().getApplicationContext();
        String basePath;

        switch (location) {
            case "documents":
                basePath = context.getFilesDir().getAbsolutePath();
                break;
            case "external":
                basePath = context.getExternalFilesDir(null).getAbsolutePath();
                break;
            default:
                basePath = context.getDatabasePath(dbName).getParent();
                break;
        }

        return basePath + File.separator + dbName;
    }

    // Helper method to copy a file from source to destination
    private boolean copyFile(File sourceFile, File destFile) throws IOException {
        try (FileInputStream fis = new FileInputStream(sourceFile);
             FileOutputStream fos = new FileOutputStream(destFile)) {
            byte[] buffer = new byte[1024];
            int length;
            while ((length = fis.read(buffer)) > 0) {
                fos.write(buffer, 0, length);
            }
            return true;
        } catch (FileNotFoundException e) {
            throw new IOException("File not found: " + e.getMessage(), e);
        }
    }

    private boolean copyDbToStorage(JSONArray args, CallbackContext callbackContext) {
        try {
            JSONObject options = args.getJSONObject(0);
            String fileName = options.getString("fileName");
            String fullPath = options.getString("fullPath");
            boolean overwrite = options.optBoolean("overwrite", false);

            // Get the app's database path for the specified file
            String sourcePath = getDatabasePath("default", fileName);

            // Destination file in storage
            File destFile = new File(fullPath);

            // Check if the destination file already exists
            if (destFile.exists()) {
                if (!overwrite) {
                    callbackContext.error("File already exists and overwrite is set to false.");
                    return false;
                } else if (!destFile.delete()) {
                    callbackContext.error("Failed to overwrite the existing file.");
                    return false;
                }
            }

            // Copy the database from sourcePath to the provided fullPath
            if (!copyFile(new File(sourcePath), destFile)) {
                callbackContext.error("Failed to copy database to storage.");
                return false;
            }

            callbackContext.success("Database copied to storage successfully.");
            return true;
        } catch (JSONException e) {
            callbackContext.error("JSON Exception: " + e.getMessage());
            return false;
        } catch (IOException e) {
            callbackContext.error("IO Exception: " + e.getMessage());
            return false;
        }
    }
}

File: src/ios/DbCopyPlugin.m
#import "DbCopyPlugin.h"

@implementation DbCopyPlugin

- (void)copyDbFromStorage:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = nil;
    @try {
        NSDictionary* options = [command.arguments objectAtIndex:0];
        NSString* dbName = options[@"dbName"];
        NSString* base64Source = options[@"base64Source"];
        NSString* location = options[@"location"] ?: @"default";
        BOOL deleteOldDb = [options[@"deleteOldDb"] boolValue];

        // Determine the destination path for the database
        NSString* destPath = [self getDatabasePath:location dbName:dbName];

        // Check if the old DB exists and if it should be deleted
        NSFileManager* fileManager = [NSFileManager defaultManager];
        if (deleteOldDb && [fileManager fileExistsAtPath:destPath]) {
            NSError* error;
            if (![fileManager removeItemAtPath:destPath error:&error]) {
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[NSString stringWithFormat:@"Failed to delete old database: %@", error.localizedDescription]];
                [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
                return;
            }
        }

        // Decode Base64 string and write it to a temporary file
        NSData* decodedData = [[NSData alloc] initWithBase64EncodedString:base64Source options:0];
        NSString* tempPath = [NSTemporaryDirectory() stringByAppendingPathComponent:dbName];
        [decodedData writeToFile:tempPath atomically:YES];

        // Copy the temporary file to the destination
        NSError* copyError;
        if (![fileManager copyItemAtPath:tempPath toPath:destPath error:&copyError]) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[NSString stringWithFormat:@"Failed to copy database: %@", copyError.localizedDescription]];
        } else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Database copied successfully."];
        }
    } @catch (NSException* exception) {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:exception.reason];
    }

    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)copyDbToStorage:(CDVInvokedUrlCommand*)command {
    CDVPluginResult* pluginResult = nil;
    @try {
        NSDictionary* options = [command.arguments objectAtIndex:0];
        NSString* fileName = options[@"fileName"];
        NSString* fullPath = options[@"fullPath"];
        BOOL overwrite = [options[@"overwrite"] boolValue];

        // Get the app's database path for the specified file
        NSString* sourcePath = [self getDatabasePath:@"default" dbName:fileName];
        NSString* destPath = [fullPath stringByAppendingPathComponent:fileName];

        NSFileManager* fileManager = [NSFileManager defaultManager];

        // Create the destination directory if it doesn't exist
        NSString* destDir = [fullPath stringByDeletingLastPathComponent];
        if (![fileManager fileExistsAtPath:destDir]) {
            NSError* createDirError;
            if (![fileManager createDirectoryAtPath:destDir withIntermediateDirectories:YES attributes:nil error:&createDirError]) {
                pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[NSString stringWithFormat:@"Failed to create destination directory: %@", createDirError.localizedDescription]];
                [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
                return;
            }
        }

        // Check if the destination file exists and handle overwrite option
        if ([fileManager fileExistsAtPath:destPath] && !overwrite) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"File already exists and overwrite is set to false."];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
            return;
        }

        // Copy the database from sourcePath to the provided fullPath
        NSError* copyError;
        if (![fileManager copyItemAtPath:sourcePath toPath:destPath error:&copyError]) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[NSString stringWithFormat:@"Failed to copy database to storage: %@", copyError.localizedDescription]];
        } else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Database copied to storage successfully."];
        }
    } @catch (NSException* exception) {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:exception.reason];
    }

    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

// Helper method to get the database path based on location
- (NSString*)getDatabasePath:(NSString*)location dbName:(NSString*)dbName {
    NSString* basePath;
    if ([location isEqualToString:@"documents"]) {
        basePath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) firstObject];
    } else if ([location isEqualToString:@"external"]) {
        basePath = [NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES) firstObject];
    } else {
        basePath = [[self.commandDelegate pathForResource:dbName] stringByDeletingLastPathComponent];
    }
    return [basePath stringByAppendingPathComponent:dbName];
}

@end
